bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/BusyIndicator.java

```diff
@@ -70,6 +70,83 @@ public static void showWhile(Display display, Runnable runnable) {
 		}
 	}
 
+	/**
+	 * If called from a {@link Display} thread, waits for the given
+	 * <code>Future</code> to complete and provides busy feedback using the
+	 * busy indicator. While waiting for completion, pending UI events are processed
+	 * to prevent UI freeze.
+	 *
+	 * If there is no {@link Display} for the current thread, the {@link Future#get()} will be executed
+	 * ignoring any {@link ExecutionException} and no busy feedback will be displayed.
+	 *
+	 * In any case, the passed {@link Future} is returned for chaining of further operations.
+	 *
+	 * @param future the {@link Future} for which busy feedback is to be shown.
+	 * @since 3.127
+	 */
+	public static void showWhile(Future<?> future) {
+		if (future == null) {
+			SWT.error(SWT.ERROR_NULL_ARGUMENT);
+		}
+		if (!future.isDone()) {
+			Display display = Display.getCurrent();
+			if (display == null) {
+				try {
+					future.get();
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+				} catch (ExecutionException e) {
+					// ignore caller might want to handle this...
+				} catch (CancellationException e) {
+					// ignore but caller might want to check afterwards
+				}
+			} else {
+				Integer busyId = setBusyCursor(display);
+				try {
+					if (future instanceof CompletionStage<?> stage) {
+						// let us wake up from sleep once the future is done
+						stage.handle((nil1, nil2) -> {
+							if (!display.isDisposed()) {
+								try {
+									display.wake();
+								} catch (SWTException e) {
+									// ignore then, this can happen due to the async nature between our check for
+									// disposed and the actual call to wake the display can be disposed
+								}
+							}
+							return null;
+						});
+					} else {
+						// for regular features we need to use a workaround, we install a timer every
+						// few ms, that should be short enough to not be noticeable by the user and long
+						// enough to not burn more CPU time than necessary
+						int wakeTime = 100;
```

Comment:
And suddenly it's OK for you to have a 100ms lag, when it was not ok to wait a few ns for OS ....?
You could also just spawn a Thread that calls Future.get() to wait for completion and wake up display there after. That would avoid polling and lag.

Reply:
> And suddenly it's OK for you to have a 100ms lag, when it was not ok to wait a few ns for OS ....?

I don't understand what you mean with "wait a few ns for OS... Under Linux (not tested for other OS) Display.sleep() can sleep several seconds (!) (about 4 - 10 in my testing) this is a noticeable lag, while 100ms are not (for the usual human) given that this is only a fallback it is for me a good compromise between

1. running as fast as possible heating the CPU
2. Adding additional (costly) resource allocations (e.g. new threads that compete with already running code)

Also it full fills all my requirements that are any UI event is handled as soon as possible (so no UI freeze) as sleep wakes up + replace the unspecified time until sleep awakes (again) with a fixed delay.

Reply:
I am talking about that you rejected Thread.yield  for not  `as fast as possible`
https://github.com/eclipse-platform/eclipse.platform.swt/pull/1266#discussion_r1642593116
instead you now sleep 100.

Reply:
It appears that as fast as possible implies the cpu burns busy waiting on the thread so one must accept a delay of some sort.  So this appears to be a compromise given the previous approach was problematic with or without a yield.  Do you have concerns about the approach itself other than it appearing as a bit of an about face?

I would imagine that if one is to show a busy cursor one expect the future to take some significant time relative to the user’s perception such that .1 is below that threshold of perception in delaying whatever follows.

Reply:
> I am talking about that you rejected Thread.yield for not `as fast as possible` [#1266 (comment)](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1266#discussion_r1642593116) instead you now sleep 100.

1. There where concerned raised (by you and @szarnekow ) to burn to much CPU time, and you even verified that neither `onSpinWait()` nor `yield()` gives (always) overall benefits. So why we need to discuss that I addressed the concerns and replaced them by either direct (thanks @szarnekow for suggest that) in case of `CompletableFuture` (maybe the most useful / 90 % case) and a trade-off that satisfies the requirements (see last item) in case someone feels the need to use it with any other type of Future (probably rare case).
2. You recently asked to "concentrate on code", so it seems not useful to argue what was said here or there and instead make comments that target the *current* implementation.
3. If you would not have picked just a very small part, you would notice that I wrote (emphasis for clarification) "we want it to drive **the event queue** (_as fast as possible_) " and that is what happens here, any UI event will *immediately* wake up the loop and make it process the event, only the caller of `showWhile` (that has actively chosen to do a busy wait) might be delayed a little bit longer (**at maximum** 100ms) in the case that no event or anything else wake up the loop (so it is not always delayed by 100ms).

> I would imagine that if one is to show a busy cursor one expect the future to take some significant time relative to the user’s perception such that .1 is below that threshold of perception in delaying whatever follows.

Correct, if one disables the code (under Linux Debian) you will see the following effect:

https://github.com/eclipse-platform/eclipse.platform.swt/assets/1331477/c8a7dd9a-b523-46fa-9de2-fa619f505aa4

so `Display#sleep()` will possibly sleep "forever" (from user pov in this case > 15 seconds!) and sadly there is no `Display#sleep(timeout) `variant available.

Reply:
> It appears that as fast as possible implies the cpu burns busy waiting on the thread 
> so one must accept a delay of some sort.

Yea a tiny tiny delay, but not 100ms. I already wrote that Thread.sleep(0) avoids burning (tested on win), which is 100ms faster. Alternatively there is the possibility to convert a plain Future to a CompletableFuture with an additional thread. - For example by using https://docs.oracle.com/javase/10/docs/api/java/util/concurrent/CompletableFuture.html#completeAsync(java.util.function.Supplier,java.util.concurrent.Executor)
That implies a context switch but also 100ms faster.

We do not have any other API that undocumented sleeps 100. If someone has n nested or repeated calls to showWhile it would wait n times 100 making processing needlessly slow. Sleep 100 may be acceptable for a single user interaction but not for a API that may be called over and over.

>so Display#sleep()

I never advocated to DISPLAY.sleep. I gave alternatives specific to the current implementation - which at least would not deadlock anymore.

Reply:
I would prefer to see this conversation move forward relative to the current actual code proposal toward a resolution and I would prefer this to take place without an analysis of who exactly said what when.  It's great that we have at least two people here who are focused on the smallest of technical details, but this is tending to lead to conflict because the two people focus on different details, sometimes seemingly at the exclusion of the other person's details. 

So I wonder about the Thread.sleep(0) thing now because it's also not 100% clear (to me) exactly what that does.  So in that light I expect a discussion that will tend to lead to all kinds of hyperlinks to facts and external discussions about what exactly it does, leaving room for endless disagreement and likely digressing far off the beaten path.

Looking at the current code I see that the display thread is doing a "normal" display event processing loop exactly like in ModalContext:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/6a53ef47-ff34-4dd7-9bed-e71e4ccd72b4)

Is there a suggestion here to replace display.sleep() with Thread.sleep(0)?  It seems to me that that might have zero effect or might cause the display thread to stop processing any immediately-arriving event for an undefined amount of time.  I'm not sure I want to see a days long debate about what Thread.sleep(0) will do on all possible JVM implementations in the known universe far into he unforeseeable future. I fear such a digression will prove futile.  So if that's not a suggestion for improvement, then I'd prefer we leave it behind.

Anyway, as I read the code this is a normal event loop keeping the display thread as alive as possible with no delays and no busy waiting.  In that "ideal" context, there is the potential that the future completes **and**, if there is no other event arriving between when it actually completes and when the timer event wakes the display from sleep,  that there will be at most a 100ms delay for the loop as whole to terminate.  This is not a delay in event processing, but simply a delay in loop termination, which is perceived by the user via the longer busy cursor.  Note that probably when we start taking about the average such delay, it's more likely to be 50ms because the future might complete at any point during the 100ms pulse time.

I assume this the delay we'd like to improve or avoid.  Unless I overlooked some important detail, this strikes me as a bit of a nit, but then no detail is too small to escape consideration and when there is room for improvement that would be a good thing.  So yes, I see there is a consideration about doing this n times.  Keep in mind though that there is another assumption here about the delay---that there are no other events waking up the display thread,---so while there is mention of a user interaction, I still wonder what exactly is that interaction that is producing n futures (nested ones?) without spanning any event queue events?  We're delving into the hypothetical now. If these are nested, then there are n timer pulses I suppose.  I don't know if these pulse at the same time or spread out in time.  Probably an implementation detail of the OS.  If the latter, then that will reduce the 100ms, average 50ms, down to a much smaller number I suppose.  More hypothetical concerns...

I can imagine that spawning an additional thread along with n **nested** calls might result in n spawned threads which sounds, at least on the surface, like a heavy solution where the overhead of the solution might be more expensive that the delay we're trying to reduce, not to mention the CPU global warning effect.  I think some action would need to be taken to avoid spawning n threads for large n. (I wonder what a single-CPU computer will do, but let's pretend I didn't wonder that out loud.)

My sense is that we are splitting hairs to look for nits.  I would personally rather not see new threads created but rather live with a very small perceived added time of future completion as the cheaper solution.

Reply:
Just to clarify, the 100ms is only if **nothing** happens that is:

1. You don't move your mouse, you don't type, you don't do anything
2. No other code invokes Display.exec / asyncExec / timerExec
3. The OS decides that for a longer time your process has not performed any repaints or alike, so it seems static and therefore throttles spurious wakeups of the UI

If one repeatably call `showWhile` each of those calls will wake up the loop and will take over. Beside that, if I really have repeated calls and I suspect them to run < 1 second don't use a busy cursor at all.

Furthermore even if I want this (for whatever reasons) it probabbly is better to use a (lighweight) `CompletableFuture` anyways instead of a heavyweight (usually backed by one dedicated thread) `Future`....

Reply:
my suggestion is to use something like
```
if (! (future instanceof CompletableFuture)) {
 future= CompletableFuture.completeAsync(future::get, executor)
}
```
to totally avoid the case where future is not a CompletableFuture.
Even creating a new thread is in average 49ms faster then the current solution. If that really creates a new Thread at all or reuses an existing thread is just a question of the Executor used.

Reply:
> Even creating a new thread is in average 49ms faster then the current solution

No it is not is just slower, see below (I commented the CompletableFuture case for this and only used the "slow" fallback):


https://github.com/eclipse-platform/eclipse.platform.swt/assets/1331477/ed63887a-adf1-40f3-84de-299b4f173ba7

So delta between completion of feature is only 3 ms, while your solution adds a constant overhead to each invocation for possibly spawning a new thread.

Also my solution always allows the caller to wrap a plain Future like you did in whatever executor wanted ...

Reply:
> my suggestion is to use something like
> 
> ```
> if (! (future instanceof CompletableFuture)) {
>  future= CompletableFuture.completeAsync(future::get, executor)
> }
> ```
> 
> to totally avoid the case where future is not a CompletableFuture. Even creating a new thread is in average 49ms faster then the current solution. If that really creates a new Thread at all or reuses an existing thread is just a question of the Executor used.

This would require passing in an appropriate executor, one that we **know** is not preoccupied with other futures that are not yet complete, and it would only be used in some cases not all cases, right?

Reply:
> This would require passing in an appropriate executor, one that we **know** is not preoccupied with other futures that are not yet complete, 

It can have running futures but must guarantee it does not reuse their threads while running. So it has to be unlimited pool size. for example Executors.html#newCachedThreadPool

> and it would only be used in some cases not all cases, right?

i do not understand that question: What cases do you distinct?

Reply:
> Also my solution always allows the caller to wrap a plain Future like you did in whatever executor wanted ...

if we restrict the API to CompletableFuture i am fine.

Reply:
> > and it would only be used in some cases not all cases, right?
> 
> i do not understand that question: What cases do you distinct?

These are two cases:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/df55a04b-139c-4522-9663-b32a276d74d6)

They would be collapsed into one case if the argument were CompletableFuture.

The proposal to use an executor in the code is less than ideal because:

- It requires another argument.
- It requires documenting the properties of executor argument.
- Alternatively something must create an executor with appropriate properties and manage it.
- It's an argument that's not used in all cases.  

On that basis, i think it presents a poor API or provides a heavy implementation.

So we seem to have two choices:

1. Change it to CompletableFuture which **maybe** makes everyone happy.
2. Accept it as proposed which makes everyone except @jukzi  happy.

I will leave the choice to @laeubi and if he chooses 2, then we will ask a team lead to resolve the impasse and move on.

Reply:
> * It requires another argument.
> * It requires documenting the properties of executor argument.
> * Alternatively something must create an executor with appropriate properties and manage it.
> * It's an argument that's not used in all cases.

That wrong. I suggested to use a Executors.html#newCachedThreadPool which does not need to be passed as argument.

Reply:
I consider the statement:

> **Alternatively** something must create an executor with appropriate properties and manage it.

and the statement

> I suggested to use a Executors.html#newCachedThreadPool which does not need to be passed as argument.

to be mutually compatible. 

Imagine instead of starting out with "that wong", you concretely suggest that the BusyIndicator could maintain a static final newCachedThreadPool and use that.   Picking up from the "Alternative..." instead of negating everything that I said as "that wron".  You could even argue that is is not so heavy weight because threads are reused, created only on demand, and are garbage collected relatively soon.  Then I would have food for thought without getting bashed in a battle of who's right and who's wrong.

When we mince every word and split every hair we will find conflict everywhere we dig deep enough. I feel like I've walked out onto a target range where a shot might come from any direction.   This is not the type of environment in which I like to work.

Reply:
As far as i understand the following test should print at max 2, and merks argued it would print ~50 in average, right?
```
	@Test
	public void testShowWhilePerformance() {
		Shell shell = new Shell();
		ExecutorService executorService = Executors.newCachedThreadPool();
		for (int i=0; i<100; i++) {
			Future<?> future = executorService.submit(() -> {
				try {
					Thread.sleep(1); //  1ms!!
				} catch (InterruptedException e) {
				}
			});
			long n0 = System.currentTimeMillis();
			BusyIndicator.showWhile(future);
			long n1 = System.currentTimeMillis();
			System.out.println((n1-n0));
		}
	}
```

Reply:
No hair is too fine.   I wonder where this is this going though.

In any case, if I run this (on Windows 10) I get it printing between 107 and 116.   If I change it to `Thread.sleep(90)`, I get effectively the same printed result.  If I change it to `Thread.sleep(100);`, sometimes it will print 200+ but mostly 100+.    Change it to `Thread.sleep(1000); ` and it prints around 1100.   Same for `Thread.sleep(1090);`.  So there is an overhead/delay that is max ~100 and min ~ 0 depending on when completion of the future falls within the wake pulse.  So yes, if you write a test case where completion always happens near the start of the pulse, the time delay will always be near the max, and vice versa, the time delay will be near the min.  Somewhere there is an average (for variable length futures) and I expect that is somewhere around the middle.  If there are other events waking up the display.sleep, the delay is reduced.  So personally, I generally expect an average of a 50ms delay/overhead on variable length futures, regardless how how long those futures are. I also personally expect that a 1ms future will generally always have roughly a 100ms overhead.  That appears to be the case here.

Reply:
> printing between 107

Exactly. And i am not concerned if it prints ~50 or ~100 - its just a showcase that the proposed implementation does have a significant overhead that depends on the configured wakeTime which could be easy avoided as you said:

> maintain a static final newCachedThreadPool and use that
> that is is not so heavy weight because threads are reused, created only on demand, and are garbage collected relatively soon.

Reply:
I'm not sure why we had demonstrate/prove exactly what I believed to be an established fact.    That's time I could have better spent differently.  

Anyway, the following proposal does not compile:
```
					ExecutorService executor = Executors.newCachedThreadPool();
					if (!(future instanceof CompletableFuture)) {
						future = CompletableFuture.completeAsync(future::get, executor);
					}
```
Perhaps the proposed code meant to be completed by the intelligent receiver.  Unfortunately I'm not that intelligent.  Google leads me to examples that I find complicated:

https://stackoverflow.com/questions/23301598/transform-java-future-into-a-completablefuture

Was the intent that someone else produce logic to convert a Future to a CompletableFuture to test/demonstrate that it's faster and better?

Reply:
yea java is complicated:
```
					ExecutorService executor = Executors.newCachedThreadPool();
					CompletableFuture<?> stage;
					if (future instanceof CompletableFuture s) {
						stage = s;
					} else {
						stage = CompletableFuture.supplyAsync(() -> {
							try {
								return future.get();
							} catch (Exception e) {
								return null;
							}
						}, executor);
					}
```

Reply:
So here is a complete implementation based on the proposal
```
	static ExecutorService executor;

	private static ExecutorService getExecutor() {
		if (executor == null) {
			executor = Executors.newCachedThreadPool();
		}
		return executor;
	}

	/**
	 * If called from a {@link Display} thread, waits for the given
	 * <code>Future</code> to complete and provides busy feedback using the busy
	 * indicator. While waiting for completion, pending UI events are processed to
	 * prevent UI freeze.
	 *
	 * If there is no {@link Display} for the current thread, the
	 * {@link Future#get()} will be executed ignoring any {@link ExecutionException}
	 * and no busy feedback will be displayed.
	 *
	 * In any case, the passed {@link Future} is returned for chaining of further
	 * operations.
	 *
	 * @param future the {@link Future} for which busy feedback is to be shown.
	 * @since 3.127
	 */
	public static void showWhile(Future<?> future) {
		if (future == null) {
			SWT.error(SWT.ERROR_NULL_ARGUMENT);
		}
		if (!future.isDone()) {
			Display display = Display.getCurrent();
			if (display == null) {
				get(future);
			} else {
				Integer busyId = setBusyCursor(display);
				try {
					CompletionStage<?> stage;
					if (future instanceof CompletionStage<?> s) {
						stage = s;
					} else {
						// convert to a stage
						stage = CompletableFuture.runAsync(() -> {
							get(future);
						}, getExecutor());
					}

					// let us wake up from sleep once the future is done
					stage.handle((nil1, nil2) -> {
						if (!display.isDisposed()) {
							try {
								display.wake();
							} catch (SWTException e) {
								// ignore then, this can happen due to the async nature between our check for
								// disposed and the actual call to wake the display can be disposed
							}
						}
						return null;
					});

					while (!future.isDone() && !display.isDisposed()) {
						if (!display.readAndDispatch()) {
							display.sleep();
						}
					}
				} finally {
					clearBusyCursor(display, busyId);
				}
			}
		}
	}

	private static void get(Future<?> future) {
		try {
			future.get();
		} catch (InterruptedException e) {
			Thread.currentThread().interrupt();
		} catch (ExecutionException e) {
			// ignore caller might want to handle this...
		} catch (CancellationException e) {
			// ignore but caller might want to check afterwards
		}
	}
```

I've avoided creating the executor until it's needed.  I've factored out the exception handling for future.get assuming that logic is important for both calls to future.get.

The test prints anything from 1 to 4.  

This looks quite simple and clean to me, and it performs well.  One could even that argue it looks cleaner and simpler.

Reply:
> Exactly. And i am not concerned if it prints ~50 or ~100 - its just a showcase that the proposed implementation does have a significant overhead

As I said below:

> As many things in SWT also this feature might be misused, but this was never some project goal of SWT that it is completely fool-proof.

This "testcase" shows exactly such "misuses", is completely synthetic and is only crafted towards sowing a "problem" that never exits in a real application:

1. Everything that application do is show nothing to the user
2. Showing a busy cursor on an trivial operation  (< 1 second expected run time)

You can trivially rewrite your example to a "super fast performing do nothing loop" like this:

```
public void testShowWhilePerformance() {
	Shell shell = new Shell();
	ExecutorService executorService = Executors.newCachedThreadPool();
	for (int i=0; i<100; i++) {
		long n0 = System.currentTimeMillis();
		BusyIndicator.execute(()->{
               	Thread.sleep(1);
		}, executorService).join();
		long n1 = System.currentTimeMillis();
		System.out.println((n1-n0));
	}
}
``` 

So my impression here is you only try to being "right" instead of either really having technical concerns...

> This looks quite simple and clean to me, and it performs well. 

Please see above, also it does *not* perform well in a real application, it performs "well" because we have a completely odd case where the application has nothing to do than spin useless threads that has nothing to do.

In a real application (where these threads will do actual useful things) you will see now two threadpools with the double size of threads each thread will usually need (depending on configuration) 1Mb per Thread we allocate another 100MB just to wait for the other 100 Threads (100MB) and all this must later be reclaimed by GC,

Beside that even the "bad example" can be rewritten to
```
Shell shell = new Shell();
	ExecutorService executorService = Executors.newCachedThreadPool();
	List<Future<?>> weNeedToWaitFor = new ArrayList<>();
	for (int i=0; i<100; i++) {
		Future<?> future = executorService.submit(() -> {
			try {
				Thread.sleep(1); //  1ms!!
			} catch (InterruptedException e) {
			}
		});
		weNeedToWaitFor.add(future);

	}
	for (int i=0; i<100; i++) {
		long n0 = System.currentTimeMillis();
		BusyIndicator.showWhile(weNeedToWaitFor.get(i));
		long n1 = System.currentTimeMillis();
		System.out.println((n1-n0));
	}
```

what shows the complete synthetic nature of this "testcase", where it always completes in zero time.

Reply:
Please reduce the timeout to 50ms (this is small enough value to not be noticable). Regarging thread pools usage - the less threads a UI toolkit creates/manages the better. Better management of threads and etc. is not a job for the UI toolkit.